gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -240,73 +178,234 @@ protected Path getPublisherOutputDir(WorkUnitState workUnitState, int branchId)
-          preserveFileName
-              ? new Path(publisherOutputDir,
-                  workUnitState.getProp(ForkOperatorUtils.getPropertyNameForBranch(
-                      ConfigurationKeys.DATA_PUBLISHER_FINAL_NAME, this.numBranches, branchId)))
-          : new Path(publisherOutputDir, status.getPath().getName());
-
-      LOG.info(String.format("Moving %s to %s", status.getPath(), finalOutputPath));
-      parallelRunner.movePath(status.getPath(), this.publisherFileSystemByBranches.get(branchId),
-          finalOutputPath, Optional.<String> absent());
-    }
+  private CompositeCommitAction getCompositeCommitAction(Iterable<PublishCommand> publishCommands) {
+    return new CompositeCommitAction(Iterables.transform(publishCommands,
+            new Function<PublishCommand, CommitAction>() {
+              @Override
+              public CommitAction apply(PublishCommand publishCommand) {
+                return publishCommand.getCommitAction();
+              }
+            }));
+  }
+
+  private static void markWorkUnitCommitted(WorkUnitState state) {
+    state.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
   }
 
-  private ParallelRunner getParallelRunner(FileSystem fs) {
-    String uri = fs.getUri().toString();
-    if (!this.parallelRunners.containsKey(uri)) {
-      this.parallelRunners.put(uri, this.closer.register(new ParallelRunner(this.parallelRunnerThreads, fs)));
+  private static class CompositeCommitAction implements Action {
+    private final Iterable<CommitAction> commitActions;
+
+    public CompositeCommitAction(Iterable<CommitAction> commitActions) {
+      this.commitActions = commitActions;
+    }
+
+    @Override
+    public void apply() throws Exception {
+      for (CommitAction commitAction : commitActions) {
+        commitAction.apply();
+      }
     }
-    return this.parallelRunners.get(uri);
   }
 
-  @Override
-  public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+  private static class CommitAction implements Action {
+    private final WorkUnitState state;
+    private AtomicInteger requiredSuccesses = new AtomicInteger();
+
+    public CommitAction(WorkUnitState state) {
+      this.state = state;
+      this.requiredSuccesses.incrementAndGet();
+    }
+
+    @Override
+    public void apply() {
+      if (requiredSuccesses.decrementAndGet() == 0) {
+        markWorkUnitCommitted(state);
+      }
+    }
+
+    public void increment() {
+      this.requiredSuccesses.incrementAndGet();
+    }
   }
 
-  @Override
-  public void publishMetadata(WorkUnitState state) throws IOException {
-    // Nothing to do
+  private static class PublishGroup {
+    private final int branchId;
+    private final FileSystem srcFs;
+    private final Path srcPath;
+    private final FileSystem dstFs;
+    private final Path dstPath;
+
+    public PublishGroup(int branchId, FileSystem srcFs, Path srcPath, FileSystem dstFs, Path dstPath) {
```

Comment:
For a given `List` of `WorkUnitState`s, when will this code ever create duplicate `PublishGroup`s? Won't the (branch, srcFs, srcPath, dstFs, dstPath) uniquely identify every `WorkUnitState`?